internal/configs/provider.go

```diff
@@ -290,3 +333,89 @@ func checkProviderNameNormalized(name string, declrange hcl.Range) hcl.Diagnosti
 	}
 	return diags
 }
+
+func (p *Provider) generateForEachProviders(eval *StaticEvaluator) ([]*Provider, hcl.Diagnostics) {
+	var diags hcl.Diagnostics
+	if eval == nil {
+		return nil, diags.Append(&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Iteration not allowed in test files",
+			Detail:   "for_each was declared as an provider attribute in a test file",
+			Subject:  p.AliasExpr.Range().Ptr(),
+		})
+	}
+
+	foeachRefsFunc := func(refs []*addrs.Reference) (*hcl.EvalContext, tfdiags.Diagnostics) {
+		var diags tfdiags.Diagnostics
+		evalContext, evalDiags := eval.EvalContext(StaticIdentifier{
+			Module:    eval.call.addr,
+			Subject:   fmt.Sprintf("provider.%s.for_each", p.Name),
+			DeclRange: p.ForEach.Range(),
+		}, refs)
+		return evalContext, diags.Append(evalDiags)
+	}
+
+	forVal, evalDiags := evalchecks.EvaluateForEachExpression(p.ForEach, foeachRefsFunc)
+	diags = append(diags, evalDiags.ToHCL()...)
+	if evalDiags.HasErrors() {
+		return nil, diags
+	}
+
+	var out []*Provider
+	for k, v := range forVal {
+		if !hclsyntax.ValidIdentifier(k) {
+			return nil, diags.Append(&hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "Invalid Identifier",
+				Detail:   fmt.Sprintf("The provided identifier %s is invalid", k),
+				Subject:  p.ForEach.Range().Ptr(),
+			})
+		}
+
+		v := v
+		iter := *p
+		iter.Alias = k
+		iter.EachValue = &v
+		iter.ForEach = nil
+		out = append(out, &iter)
+	}
+	return out, diags
+}
+
+func (p *Provider) generateCountProviders(eval *StaticEvaluator) ([]*Provider, hcl.Diagnostics) {
+	var diags hcl.Diagnostics
+	if eval == nil {
+		return nil, diags.Append(&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Iteration not allowed in test files",
+			Detail:   "for_each was declared as an provider attribute in a test file",
+			Subject:  p.AliasExpr.Range().Ptr(),
+		})
+	}
+	countEvalFunc := func(expr hcl.Expression) (cty.Value, tfdiags.Diagnostics) {
+		var diags tfdiags.Diagnostics
+		countVal, evalDiags := eval.Evaluate(expr, StaticIdentifier{
+			Module:    eval.call.addr,
+			Subject:   fmt.Sprintf("provider.%s.count", p.Name),
+			DeclRange: p.Count.Range(),
+		})
+		return countVal, diags.Append(evalDiags)
+	}
+
+	var out []*Provider
+	countVal, evalDiags := evalchecks.EvaluateCountExpression(p.Count, countEvalFunc)
+	diags = append(diags, evalDiags.ToHCL()...)
+	if evalDiags.HasErrors() {
+		return nil, diags
+	}
+
+	for i := 0; i < countVal; i++ {
+		iter := *p
+		iter.Alias = fmt.Sprintf("[%d]", i)
```

Comment:
Do we want count providers to be available only via square brackets? Should we let the caller decide?

Reply:
interesting. so i had made changes to resource.go so i could reference a provider like so: 

```
provider = local.0
```

this made use of https://github.com/opentofu/opentofu/blob/main/internal/addrs/instance_key.go#L66-L68 

this meant that it had to be wrapped in square brackets to match. However thats all been remove as it steps on the toes of what you are working on. So not sure if i still need them 🤔